segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMapper.java

```diff
@@ -180,14 +177,88 @@ public StreamSegmentMapper(ContainerMetadata containerMetadata, OperationLog dur
-                .thenComposeAsync(transInfo -> this.stateStore.put(transInfo.getName(), getState(transInfo, attributes), timer.getRemaining()), this.executor)
+                .thenComposeAsync(parentId -> createSegmentInStorageWithRecovery(transactionName, attributes, timer), this.executor)
                 .thenApply(v -> {
                     LoggerHelpers.traceLeave(log, traceObjectId, "createNewTransactionStreamSegment", traceId, parentStreamSegmentName, transactionName);
                     return transactionName;
                 });
     }
 
+    /**
+     * Attempts to create the given Segment (or Transaction) in Storage, with possible recovery from a previous incomplete
+     * attempt. When this method completes successfully, the Storage will have the Segment created, as well as a State File
+     * with the appropriate contents.
+     * <p>
+     * The recovery part handles these three major cases:
+     * <ul>
+     * <li>Segment exists and has a valid State File: the operation will fail with StreamSegmentExistsException.
+     * <li>Segment exists, has a length of zero, and a missing or invalid State File: the state file will be recreated using
+     * the given attributes and the operation will complete successfully (pending a successful State File creation).
+     * <li>Segment exists, has a non-zero length, and either a valid or invalid/missing State File: the state file will be
+     * recreated (if needed) using the given attributes and the operation will fail with StreamSegmentExistsException.
+     * </ul>
+     *
+     * @param segmentName The name of the Segment/Transaction to create.
+     * @param attributes  The initial Attributes for the Segment, if any.
+     * @param timer       A TimeoutTimer to determine how much time is left to complete the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the Segment has been successfully created.
+     */
+    private CompletableFuture<Void> createSegmentInStorageWithRecovery(String segmentName, Collection<AttributeUpdate> attributes, TimeoutTimer timer) {
+        return FutureHelpers
+                .exceptionallyCompose(
+                        this.storage.create(segmentName, timer.getRemaining()),
+                        ex -> handleStorageCreateException(segmentName, ExceptionHelpers.getRealException(ex), timer))
+                .thenComposeAsync(si ->
+                                // Need to create the state file before we throw any further exceptions in order to recover from
+                                // previous partial executions (where we created a segment but no or empty state file).
+                                this.stateStore.put(segmentName, getState(si, attributes), timer.getRemaining())
+                                               .thenRun(() -> {
+                                                   if (si.getLength() > 0) {
```

Comment:
I need some clarification in the logic here:

1- Shouldn't we check whether the length is greater than zero before the `put` call? The `put` call will write the state file, no?
2- In the case the initial length is greater than zero, aren't we supposed to check whether the content is correct by reading and deserializing it? I saw that if the deserialization fails, then we throw data corruption exception.

Reply:
I think I understand what this is trying to do now and have answered my own questions. The put will write the state file independent of it existing or not and its content. The verification actually happens in `handleStorageCreateException`. 

The length check is still puzzling me a bit, though. I think this is trying to distinguish between the case in which the state file has been partially populated or not populated at all, and in the case it has been partially populated, we return that the segment already exist. If the previous attempt did not complete, then it sounds better this time around to return success rather than segment already exists. what do you think?

Reply:
Actually, from the test cases, it looks like we are not returning segment exists in the case of empty/corrupted state file, so I'm still missing the point of this length check for throwing segment exists.

Reply:
The contract says that if we return `StreamSegmentExistsException`, then the segment must exist and must be in a usable state. In our case, if we get in here, the segment file does exist, but the state file does not or is corrupted. We need to fix it before returning so that the Controller is assured the segment can be used.